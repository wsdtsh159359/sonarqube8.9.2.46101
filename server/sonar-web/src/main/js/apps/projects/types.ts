/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { ComponentQualifier } from '../../types/component';

export interface Project {
  analysisDate?: string;
  isFavorite?: boolean;
  key: string;
  leakPeriodDate?: string;
  measures: T.Dict<string>;
  name: string;
  projects?: number;
  qualifier: ComponentQualifier;
  tags: string[];
  visibility: T.Visibility;
  needIssueSync?: boolean;
}

export interface Facet {
  [value: string]: number;
}

export interface Facets {
  [property: string]: Facet;
}
