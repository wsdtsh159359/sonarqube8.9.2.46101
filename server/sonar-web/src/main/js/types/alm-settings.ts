/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
export const enum AlmKeys {
  Azure = 'azure',
  BitbucketServer = 'bitbucket',
  BitbucketCloud = 'bitbucketcloud',
  GitHub = 'github',
  GitLab = 'gitlab'
}

export interface AlmBindingDefinition {
  key: string;
  url?: string;
}

export interface AzureBindingDefinition extends AlmBindingDefinition {
  personalAccessToken: string;
  url?: string;
}

export interface BitbucketBindingDefinition extends AlmBindingDefinition {
  personalAccessToken: string;
  url: string;
}

export interface BitbucketCloudBindingDefinition extends AlmBindingDefinition {
  clientId: string;
  clientSecret: string;
  workspace: string;
}

export interface GithubBindingDefinition extends AlmBindingDefinition {
  appId: string;
  clientId: string;
  clientSecret: string;
  privateKey: string;
  url: string;
}

export interface GitlabBindingDefinition extends AlmBindingDefinition {
  personalAccessToken: string;
  url?: string;
}

export interface ProjectAlmBindingResponse {
  alm: AlmKeys;
  key: string;
  repository?: string;
  slug?: string;
  summaryCommentEnabled?: boolean;
  monorepo: boolean;
}

export interface ProjectAzureBindingResponse extends ProjectAlmBindingResponse {
  alm: AlmKeys.Azure;
  repository: string;
  slug: string;
  url: string;
}

export interface ProjectBitbucketBindingResponse extends ProjectAlmBindingResponse {
  alm: AlmKeys.BitbucketServer;
  repository: string;
  slug: string;
}

export interface ProjectGitHubBindingResponse extends ProjectAlmBindingResponse {
  alm: AlmKeys.GitHub;
  repository: string;
}

export interface ProjectGitLabBindingResponse extends ProjectAlmBindingResponse {
  alm: AlmKeys.GitLab;
  repository: string;
  url: string;
}

export interface ProjectAlmBindingParams {
  almSetting: string;
  project: string;
  monorepo: boolean;
}

export interface AzureProjectAlmBindingParams extends ProjectAlmBindingParams {
  projectName: string;
  repositoryName: string;
}

export interface BitbucketProjectAlmBindingParams extends ProjectAlmBindingParams {
  repository: string;
  slug: string;
}

export interface BitbucketCloudProjectAlmBindingParams extends ProjectAlmBindingParams {
  repository: string;
}

export interface GithubProjectAlmBindingParams extends ProjectAlmBindingParams {
  repository: string;
  summaryCommentEnabled: boolean;
}

export interface GitlabProjectAlmBindingParams extends ProjectAlmBindingParams {
  repository?: string;
}

export interface AlmSettingsInstance {
  alm: AlmKeys;
  key: string;
  url?: string;
}

export interface AlmSettingsBindingDefinitions {
  [AlmKeys.Azure]: AzureBindingDefinition[];
  [AlmKeys.BitbucketServer]: BitbucketBindingDefinition[];
  [AlmKeys.BitbucketCloud]: BitbucketCloudBindingDefinition[];
  [AlmKeys.GitHub]: GithubBindingDefinition[];
  [AlmKeys.GitLab]: GitlabBindingDefinition[];
}

export interface AlmSettingsBindingStatus {
  alertSuccess: boolean;
  failureMessage: string;
  type: AlmSettingsBindingStatusType;
}

export enum AlmSettingsBindingStatusType {
  Validating,
  Success,
  Failure,
  Warning
}

export function isProjectBitbucketBindingResponse(
  binding: ProjectAlmBindingResponse
): binding is ProjectBitbucketBindingResponse {
  return binding.alm === AlmKeys.BitbucketServer;
}

export function isProjectGitHubBindingResponse(
  binding: ProjectAlmBindingResponse
): binding is ProjectGitHubBindingResponse {
  return binding.alm === AlmKeys.GitHub;
}

export function isProjectGitLabBindingResponse(
  binding: ProjectAlmBindingResponse
): binding is ProjectGitLabBindingResponse {
  return binding.alm === AlmKeys.GitLab;
}

export function isProjectAzureBindingResponse(
  binding: ProjectAlmBindingResponse
): binding is ProjectAzureBindingResponse {
  return binding.alm === AlmKeys.Azure;
}

export function isBitbucketBindingDefinition(
  binding?: AlmBindingDefinition & { url?: string }
): binding is BitbucketBindingDefinition {
  return binding !== undefined && binding.url !== undefined;
}

export function isBitbucketCloudBindingDefinition(
  binding?: AlmBindingDefinition & { clientId?: string; workspace?: string }
): binding is BitbucketCloudBindingDefinition {
  return binding !== undefined && binding.clientId !== undefined && binding.workspace !== undefined;
}

export function isGithubBindingDefinition(
  binding?: AlmBindingDefinition & { appId?: string; url?: string }
): binding is GithubBindingDefinition {
  return binding !== undefined && binding.appId !== undefined && binding.url !== undefined;
}

export function isGitLabBindingDefinition(
  binding?: AlmBindingDefinition | GithubBindingDefinition | BitbucketCloudBindingDefinition
): binding is GitlabBindingDefinition {
  // There's too much overlap with the others. We must not only validate that certain fields are
  // present, we must also validate that others are NOT present. And even so, we cannot be 100%
  // sure, as right now, Azure, Bitbucket Server, and GitLab have the same signature.
  return (
    binding !== undefined &&
    (binding as GithubBindingDefinition).appId === undefined &&
    (binding as BitbucketCloudBindingDefinition).workspace === undefined
  );
}
