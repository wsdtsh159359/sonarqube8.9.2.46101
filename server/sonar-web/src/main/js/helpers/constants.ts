/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { colors } from '../app/theme';
import { AlmKeys } from '../types/alm-settings';
import { ComponentQualifier } from '../types/component';
import { IssueScope, IssueType } from '../types/issues';

export const SEVERITIES = ['BLOCKER', 'CRITICAL', 'MAJOR', 'MINOR', 'INFO'];
export const STATUSES = ['OPEN', 'REOPENED', 'CONFIRMED', 'RESOLVED', 'CLOSED'];
export const ISSUE_TYPES: T.IssueType[] = [
  IssueType.Bug,
  IssueType.Vulnerability,
  IssueType.CodeSmell,
  IssueType.SecurityHotspot
];
export const SOURCE_SCOPES = [
  { scope: IssueScope.Main, qualifier: ComponentQualifier.File },
  { scope: IssueScope.Test, qualifier: ComponentQualifier.TestFile }
];
export const RULE_TYPES: T.RuleType[] = ['BUG', 'VULNERABILITY', 'CODE_SMELL', 'SECURITY_HOTSPOT'];
export const RULE_STATUSES = ['READY', 'BETA', 'DEPRECATED'];

export const CHART_COLORS_RANGE_PERCENT = [
  colors.green,
  colors.lightGreen,
  colors.yellow,
  colors.orange,
  colors.red
];

export const CHART_REVERSED_COLORS_RANGE_PERCENT = [
  colors.red,
  colors.orange,
  colors.yellow,
  colors.lightGreen,
  colors.green
];

export const RATING_COLORS = [
  colors.green,
  colors.lightGreen,
  colors.yellow,
  colors.orange,
  colors.red
];

export const PROJECT_KEY_MAX_LEN = 400;

export const ALM_DOCUMENTATION_PATHS = {
  [AlmKeys.Azure]: '/documentation/analysis/azuredevops-integration/',
  [AlmKeys.BitbucketServer]: '/documentation/analysis/bitbucket-integration/',
  [AlmKeys.BitbucketCloud]: '/documentation/analysis/bitbucket-cloud-integration/',
  [AlmKeys.GitHub]: '/documentation/analysis/github-integration/',
  [AlmKeys.GitLab]: '/documentation/analysis/gitlab-integration/'
};

export const IMPORT_COMPATIBLE_ALMS = [
  AlmKeys.Azure,
  AlmKeys.BitbucketServer,
  AlmKeys.GitHub,
  AlmKeys.GitLab
];
