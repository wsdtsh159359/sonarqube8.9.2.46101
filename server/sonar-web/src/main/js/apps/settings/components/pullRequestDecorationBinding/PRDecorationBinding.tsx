/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { connect } from 'react-redux';
import { HttpStatus } from 'sonar-ui-common/helpers/request';
import {
  deleteProjectAlmBinding,
  getAlmSettings,
  getProjectAlmBinding,
  setProjectAzureBinding,
  setProjectBitbucketBinding,
  setProjectBitbucketCloudBinding,
  setProjectGithubBinding,
  setProjectGitlabBinding
} from '../../../../api/alm-settings';
import throwGlobalError from '../../../../app/utils/throwGlobalError';
import { getAppState, Store } from '../../../../store/rootReducer';
import {
  AlmKeys,
  AlmSettingsInstance,
  ProjectAlmBindingResponse
} from '../../../../types/alm-settings';
import { EditionKey } from '../../../../types/editions';
import PRDecorationBindingRenderer from './PRDecorationBindingRenderer';

type FormData = T.Omit<ProjectAlmBindingResponse, 'alm'>;

interface StateProps {
  monorepoEnabled: boolean;
}

interface Props {
  component: T.Component;
}

interface State {
  formData: FormData;
  instances: AlmSettingsInstance[];
  isChanged: boolean;
  isConfigured: boolean;
  isValid: boolean;
  loading: boolean;
  orignalData?: FormData;
  saving: boolean;
  success: boolean;
}

const REQUIRED_FIELDS_BY_ALM: {
  [almKey in AlmKeys]: Array<keyof T.Omit<FormData, 'key'>>;
} = {
  [AlmKeys.Azure]: ['repository', 'slug'],
  [AlmKeys.BitbucketServer]: ['repository', 'slug'],
  [AlmKeys.BitbucketCloud]: ['repository'],
  [AlmKeys.GitHub]: ['repository'],
  [AlmKeys.GitLab]: ['repository']
};

export class PRDecorationBinding extends React.PureComponent<Props & StateProps, State> {
  mounted = false;
  state: State = {
    formData: { key: '', monorepo: false },
    instances: [],
    isChanged: false,
    isConfigured: false,
    isValid: false,
    loading: true,
    saving: false,
    success: false
  };

  componentDidMount() {
    this.mounted = true;
    this.fetchDefinitions();
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  fetchDefinitions = () => {
    const project = this.props.component.key;
    return Promise.all([getAlmSettings(project), this.getProjectBinding(project)])
      .then(([instances, originalData]) => {
        if (this.mounted) {
          this.setState(({ formData }) => {
            const newFormData = originalData || formData;
            return {
              formData: newFormData,
              instances: instances || [],
              isChanged: false,
              isConfigured: !!originalData,
              isValid: this.validateForm(newFormData),
              loading: false,
              orignalData: newFormData
            };
          });
        }
      })
      .catch(() => {
        if (this.mounted) {
          this.setState({ loading: false });
        }
      });
  };

  getProjectBinding(project: string): Promise<ProjectAlmBindingResponse | undefined> {
    return getProjectAlmBinding(project).catch((response: Response) => {
      if (response && response.status === HttpStatus.NotFound) {
        return undefined;
      }
      return throwGlobalError(response);
    });
  }

  catchError = () => {
    if (this.mounted) {
      this.setState({ saving: false });
    }
  };

  handleReset = () => {
    const { component } = this.props;
    this.setState({ saving: true });
    deleteProjectAlmBinding(component.key)
      .then(() => {
        if (this.mounted) {
          this.setState({
            formData: {
              key: '',
              repository: '',
              slug: '',
              monorepo: false
            },
            orignalData: undefined,
            isChanged: false,
            isConfigured: false,
            saving: false,
            success: true
          });
        }
      })
      .catch(this.catchError);
  };

  submitProjectAlmBinding(
    alm: AlmKeys,
    key: string,
    almSpecificFields?: T.Omit<FormData, 'key'>
  ): Promise<void> {
    const almSetting = key;
    const project = this.props.component.key;
    const repository = almSpecificFields?.repository;
    const slug = almSpecificFields?.slug;
    const monorepo = almSpecificFields?.monorepo ?? false;

    if (!repository) {
      return Promise.reject();
    }

    switch (alm) {
      case AlmKeys.Azure: {
        if (!slug) {
          return Promise.reject();
        }
        return setProjectAzureBinding({
          almSetting,
          project,
          projectName: slug,
          repositoryName: repository,
          monorepo
        });
      }
      case AlmKeys.BitbucketServer: {
        if (!slug) {
          return Promise.reject();
        }
        return setProjectBitbucketBinding({
          almSetting,
          project,
          repository,
          slug,
          monorepo
        });
      }
      case AlmKeys.BitbucketCloud: {
        return setProjectBitbucketCloudBinding({
          almSetting,
          project,
          repository,
          monorepo
        });
      }
      case AlmKeys.GitHub: {
        // By default it must remain true.
        const summaryCommentEnabled = almSpecificFields?.summaryCommentEnabled ?? true;
        return setProjectGithubBinding({
          almSetting,
          project,
          repository,
          summaryCommentEnabled,
          monorepo
        });
      }

      case AlmKeys.GitLab: {
        return setProjectGitlabBinding({
          almSetting,
          project,
          repository,
          monorepo
        });
      }

      default:
        return Promise.reject();
    }
  }

  handleSubmit = () => {
    this.setState({ saving: true });
    const {
      formData: { key, ...additionalFields },
      instances
    } = this.state;

    const selected = instances.find(i => i.key === key);
    if (!key || !selected) {
      return;
    }

    this.submitProjectAlmBinding(selected.alm, key, additionalFields)
      .then(() => {
        if (this.mounted) {
          this.setState({
            saving: false,
            success: true
          });
        }
      })
      .then(this.fetchDefinitions)
      .catch(this.catchError);
  };

  isDataSame(
    { key, repository = '', slug = '', summaryCommentEnabled = false, monorepo = false }: FormData,
    {
      key: oKey = '',
      repository: oRepository = '',
      slug: oSlug = '',
      summaryCommentEnabled: osummaryCommentEnabled = false,
      monorepo: omonorepo = false
    }: FormData
  ) {
    return (
      key === oKey &&
      repository === oRepository &&
      slug === oSlug &&
      summaryCommentEnabled === osummaryCommentEnabled &&
      monorepo === omonorepo
    );
  }

  handleFieldChange = (id: keyof ProjectAlmBindingResponse, value: string | boolean) => {
    this.setState(({ formData, orignalData }) => {
      const newFormData = {
        ...formData,
        [id]: value
      };

      return {
        formData: newFormData,
        isValid: this.validateForm(newFormData),
        isChanged: !this.isDataSame(newFormData, orignalData || { key: '', monorepo: false }),
        success: false
      };
    });
  };

  validateForm = ({ key, ...additionalFields }: State['formData']) => {
    const { instances } = this.state;
    const selected = instances.find(i => i.key === key);
    if (!key || !selected) {
      return false;
    }
    return REQUIRED_FIELDS_BY_ALM[selected.alm].reduce(
      (result: boolean, field) => result && Boolean(additionalFields[field]),
      true
    );
  };

  render() {
    const { monorepoEnabled } = this.props;

    return (
      <PRDecorationBindingRenderer
        onFieldChange={this.handleFieldChange}
        onReset={this.handleReset}
        onSubmit={this.handleSubmit}
        monorepoEnabled={monorepoEnabled}
        {...this.state}
      />
    );
  }
}

const mapStateToProps = (state: Store): StateProps => ({
  // This feature trigger will be replaced when SONAR-14349 is implemented
  monorepoEnabled: [EditionKey.enterprise, EditionKey.datacenter].includes(
    getAppState(state).edition as EditionKey
  )
});

export default connect(mapStateToProps)(PRDecorationBinding);
