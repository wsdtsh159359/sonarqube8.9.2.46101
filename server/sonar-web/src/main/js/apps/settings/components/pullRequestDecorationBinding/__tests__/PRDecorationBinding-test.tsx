/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { waitAndUpdate } from 'sonar-ui-common/helpers/testUtils';
import {
  deleteProjectAlmBinding,
  getAlmSettings,
  getProjectAlmBinding,
  setProjectAzureBinding,
  setProjectBitbucketBinding,
  setProjectBitbucketCloudBinding,
  setProjectGithubBinding,
  setProjectGitlabBinding
} from '../../../../../api/alm-settings';
import { mockComponent } from '../../../../../helpers/testMocks';
import { AlmKeys, AlmSettingsInstance } from '../../../../../types/alm-settings';
import { PRDecorationBinding } from '../PRDecorationBinding';

jest.mock('../../../../../api/alm-settings', () => ({
  getAlmSettings: jest.fn().mockResolvedValue([]),
  getProjectAlmBinding: jest.fn().mockResolvedValue(undefined),
  setProjectAzureBinding: jest.fn().mockResolvedValue(undefined),
  setProjectBitbucketBinding: jest.fn().mockResolvedValue(undefined),
  setProjectGithubBinding: jest.fn().mockResolvedValue(undefined),
  setProjectGitlabBinding: jest.fn().mockResolvedValue(undefined),
  setProjectBitbucketCloudBinding: jest.fn().mockResolvedValue(undefined),
  deleteProjectAlmBinding: jest.fn().mockResolvedValue(undefined)
}));

const PROJECT_KEY = 'project-key';

beforeEach(() => {
  jest.clearAllMocks();
});

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot();
});

it('should fill selects and fill formdata', async () => {
  const url = 'github.com';
  const instances = [{ key: 'instance1', url, alm: AlmKeys.GitHub }];
  const formdata = {
    key: 'instance1',
    repository: 'account/repo'
  };
  (getAlmSettings as jest.Mock).mockResolvedValueOnce(instances);
  (getProjectAlmBinding as jest.Mock).mockResolvedValueOnce(formdata);

  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  expect(wrapper.state().loading).toBe(false);
  expect(wrapper.state().formData).toEqual(formdata);
  expect(wrapper.state().isChanged).toBe(false);
});

it('should handle reset', async () => {
  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);
  wrapper.setState({
    formData: {
      key: 'whatever',
      repository: 'something/else',
      monorepo: false
    }
  });

  wrapper.instance().handleReset();
  await waitAndUpdate(wrapper);

  expect(deleteProjectAlmBinding).toBeCalledWith(PROJECT_KEY);
  expect(wrapper.state().formData).toEqual({ key: '', repository: '', slug: '', monorepo: false });
  expect(wrapper.state().isChanged).toBe(false);
});

describe('handleSubmit', () => {
  const instances: AlmSettingsInstance[] = [
    { key: 'github', alm: AlmKeys.GitHub },
    { key: 'azure', alm: AlmKeys.Azure },
    { key: 'bitbucket', alm: AlmKeys.BitbucketServer },
    { key: 'gitlab', alm: AlmKeys.GitLab },
    { key: 'bitbucketcloud', alm: AlmKeys.BitbucketCloud }
  ];

  it('should work for github', async () => {
    const wrapper = shallowRender();
    await waitAndUpdate(wrapper);
    const githubKey = 'github';
    const repository = 'repo/path';
    const summaryCommentEnabled = true;
    const monorepo = true;
    wrapper.setState({
      formData: { key: githubKey, repository, summaryCommentEnabled, monorepo },
      instances
    });
    wrapper.instance().handleSubmit();
    await waitAndUpdate(wrapper);

    expect(setProjectGithubBinding).toBeCalledWith({
      almSetting: githubKey,
      project: PROJECT_KEY,
      repository,
      summaryCommentEnabled,
      monorepo
    });
    expect(wrapper.state().success).toBe(true);
  });

  it('should work for azure', async () => {
    const wrapper = shallowRender();
    await waitAndUpdate(wrapper);
    const azureKey = 'azure';
    const repository = 'az-rep';
    const slug = 'az-project';
    const monorepo = true;
    wrapper.setState({
      formData: { key: azureKey, repository, slug, monorepo },
      instances
    });
    wrapper.instance().handleSubmit();
    await waitAndUpdate(wrapper);

    expect(setProjectAzureBinding).toBeCalledWith({
      almSetting: azureKey,
      project: PROJECT_KEY,
      projectName: slug,
      repositoryName: repository,
      monorepo
    });
    expect(wrapper.state().success).toBe(true);
  });

  it('should work for bitbucket', async () => {
    const wrapper = shallowRender();
    await waitAndUpdate(wrapper);
    const bitbucketKey = 'bitbucket';
    const repository = 'repoKey';
    const slug = 'repoSlug';
    const monorepo = true;
    wrapper.setState({ formData: { key: bitbucketKey, repository, slug, monorepo }, instances });
    wrapper.instance().handleSubmit();
    await waitAndUpdate(wrapper);

    expect(setProjectBitbucketBinding).toBeCalledWith({
      almSetting: bitbucketKey,
      project: PROJECT_KEY,
      repository,
      slug,
      monorepo
    });
    expect(wrapper.state().success).toBe(true);
  });

  it('should work for gitlab', async () => {
    const wrapper = shallowRender();
    await waitAndUpdate(wrapper);
    const gitlabKey = 'gitlab';
    const repository = 'repo';
    const monorepo = true;
    wrapper.setState({
      formData: { key: gitlabKey, repository, monorepo },
      instances
    });
    wrapper.instance().handleSubmit();
    await waitAndUpdate(wrapper);

    expect(setProjectGitlabBinding).toBeCalledWith({
      almSetting: gitlabKey,
      project: PROJECT_KEY,
      repository,
      monorepo
    });
    expect(wrapper.state().success).toBe(true);
  });

  it('should work for bitbucket cloud', async () => {
    const wrapper = shallowRender();
    await waitAndUpdate(wrapper);
    const bitbucketKey = 'bitbucketcloud';
    const repository = 'repoKey';
    const monorepo = true;
    wrapper.setState({ formData: { key: bitbucketKey, repository, monorepo }, instances: [] });
    wrapper.instance().handleSubmit();
    await waitAndUpdate(wrapper);
    expect(setProjectBitbucketCloudBinding).not.toBeCalled();

    wrapper.setState({ formData: { key: bitbucketKey, repository, monorepo }, instances });
    wrapper.instance().handleSubmit();
    await waitAndUpdate(wrapper);

    expect(setProjectBitbucketCloudBinding).toBeCalledWith({
      almSetting: bitbucketKey,
      project: PROJECT_KEY,
      repository,
      monorepo
    });
    expect(wrapper.state().success).toBe(true);
  });
});

describe.each([[500], [404]])('For status %i', status => {
  it('should handle failures gracefully', async () => {
    const newFormData = {
      key: 'whatever',
      repository: 'something/else',
      monorepo: false
    };

    (getProjectAlmBinding as jest.Mock).mockRejectedValueOnce({ status });
    (setProjectGithubBinding as jest.Mock).mockRejectedValueOnce({ status });
    (deleteProjectAlmBinding as jest.Mock).mockRejectedValueOnce({ status });

    const wrapper = shallowRender();
    await waitAndUpdate(wrapper);
    wrapper.setState({
      formData: newFormData,
      orignalData: undefined
    });

    wrapper.instance().handleSubmit();
    await waitAndUpdate(wrapper);
    expect(wrapper.instance().state.orignalData).toBeUndefined();
    wrapper.instance().handleReset();
    await waitAndUpdate(wrapper);
    expect(wrapper.instance().state.formData).toEqual(newFormData);
  });
});

it('should handle field changes', async () => {
  const url = 'git.enterprise.com';
  const repository = 'my/repo';
  const instances: AlmSettingsInstance[] = [
    { key: 'instance1', url, alm: AlmKeys.GitHub },
    { key: 'instance2', url, alm: AlmKeys.GitHub },
    { key: 'instance3', url: 'otherurl', alm: AlmKeys.GitHub }
  ];
  (getAlmSettings as jest.Mock).mockResolvedValueOnce(instances);
  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  wrapper.instance().handleFieldChange('key', 'instance2');
  await waitAndUpdate(wrapper);
  expect(wrapper.state().formData).toEqual({
    key: 'instance2',
    monorepo: false
  });

  wrapper.instance().handleFieldChange('repository', repository);
  await waitAndUpdate(wrapper);
  expect(wrapper.state().formData).toEqual({
    key: 'instance2',
    repository,
    monorepo: false
  });

  wrapper.instance().handleFieldChange('summaryCommentEnabled', true);
  await waitAndUpdate(wrapper);
  expect(wrapper.state().formData).toEqual({
    key: 'instance2',
    monorepo: false,
    repository,
    summaryCommentEnabled: true
  });

  wrapper.instance().handleFieldChange('monorepo', true);
  await waitAndUpdate(wrapper);
  expect(wrapper.state().formData).toEqual({
    key: 'instance2',
    repository,
    summaryCommentEnabled: true,
    monorepo: true
  });
});

it.each([
  [AlmKeys.Azure, { monorepo: false }],
  [AlmKeys.Azure, { slug: 'test', monorepo: false }],
  [AlmKeys.Azure, { repository: 'test', monorepo: false }],
  [AlmKeys.BitbucketServer, { monorepo: false }],
  [AlmKeys.BitbucketServer, { slug: 'test', monorepo: false }],
  [AlmKeys.BitbucketServer, { repository: 'test', monorepo: false }],
  [AlmKeys.BitbucketCloud, { monorepo: false }],
  [AlmKeys.GitHub, { monorepo: false }],
  [AlmKeys.GitLab, { monorepo: false }]
])(
  'should properly reject promise for %s & %s',
  async (almKey: AlmKeys, params: { monorepo: boolean }) => {
    const wrapper = shallowRender();

    expect.assertions(1);
    await expect(
      wrapper.instance().submitProjectAlmBinding(almKey, 'binding', params)
    ).rejects.toBeUndefined();
  }
);

it('should validate form', async () => {
  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  const validateMethod = wrapper.instance().validateForm;

  expect(validateMethod({ key: '', repository: '', monorepo: false })).toBe(false);
  expect(validateMethod({ key: '', repository: 'c', monorepo: false })).toBe(false);

  wrapper.setState({
    instances: [
      { key: 'azure', alm: AlmKeys.Azure },
      { key: 'bitbucket', alm: AlmKeys.BitbucketServer },
      { key: 'bitbucketcloud', alm: AlmKeys.BitbucketCloud },
      { key: 'github', alm: AlmKeys.GitHub },
      { key: 'gitlab', alm: AlmKeys.GitLab }
    ]
  });

  [
    { values: { key: 'azure', monorepo: false, repository: 'rep' }, result: false },
    { values: { key: 'azure', monorepo: false, slug: 'project' }, result: false },
    {
      values: { key: 'azure', monorepo: false, repository: 'repo', slug: 'project' },
      result: true
    },
    { values: { key: 'github', monorepo: false, repository: '' }, result: false },
    { values: { key: 'github', monorepo: false, repository: 'asdf' }, result: true },
    { values: { key: 'bitbucket', monorepo: false, repository: 'key' }, result: false },
    {
      values: { key: 'bitbucket', monorepo: false, repository: 'key', slug: 'slug' },
      result: true
    },
    { values: { key: 'bitbucketcloud', monorepo: false, repository: '' }, result: false },
    { values: { key: 'bitbucketcloud', monorepo: false, repository: 'key' }, result: true },
    { values: { key: 'gitlab', monorepo: false }, result: false },
    { values: { key: 'gitlab', monorepo: false, repository: 'key' }, result: true }
  ].forEach(({ values, result }) => {
    expect(validateMethod(values)).toBe(result);
  });
});

function shallowRender(props: Partial<PRDecorationBinding['props']> = {}) {
  return shallow<PRDecorationBinding>(
    <PRDecorationBinding
      component={mockComponent({ key: PROJECT_KEY })}
      monorepoEnabled={false}
      {...props}
    />
  );
}
