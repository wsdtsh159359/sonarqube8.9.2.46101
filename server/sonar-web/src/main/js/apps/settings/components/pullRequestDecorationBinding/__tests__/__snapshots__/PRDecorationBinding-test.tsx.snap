// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<PRDecorationBindingRenderer
  formData={
    Object {
      "key": "",
      "monorepo": false,
    }
  }
  instances={Array []}
  isChanged={false}
  isConfigured={false}
  isValid={false}
  loading={true}
  monorepoEnabled={false}
  onFieldChange={[Function]}
  onReset={[Function]}
  onSubmit={[Function]}
  saving={false}
  success={false}
/>
`;
